src/deploy/functions/checkIam.ts

```diff
@@ -51,15 +52,19 @@ export async function checkServiceAccountIam(projectId: string): Promise<void> {
  * @param options The command-wide options object.
  * @param payload The deploy payload.
  */
-export async function checkHttpIam(context: any, options: any, payload: any): Promise<void> {
-  const functionsInfo = payload.functions.triggers;
+export async function checkHttpIam(
```

Comment:
I have mixed feelings. I've gotten used to "context' being passed around in Go and I prefer editing Context as the only side-effect allowed in our API calls. I've got no love for "Options' though since the actual command line flags passed seem like too low level a concern for this code (e.g. options' project vs context's project). And I have yet to find the purpose of Payload.

Reply:
Agree with you on options and  (especially) payload.  We use this same "context, options, paylod" pattern for all of the deploy code paths, and payload might get used more in the other deploy modules (hosting, storage rules, etc), but its basically unused here. 

For context, I've bumped into a bunch of subtle bugs over my time with this codebase caused by using context as a mutable "global" state. I also find that it can make the code hard to follow sometimes - its not always obvious where context.blah was originally set or if it has been set at all. Having said that, I do think that your proposed changes (stronger typing, and moving complex objects like 'existingFunctions' out of context) alleviate these concerns a bit.